connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperator.java

```diff
@@ -0,0 +1,331 @@
+
+    private static final Logger log = LoggerFactory.getLogger(RetryWithToleranceOperator.class);
+
+    public static final String RETRY_TIMEOUT = "retry.timeout";
+    public static final String RETRY_TIMEOUT_DOC = "The maximum number of retries before failing an operation";
+    public static final long RETRY_TIMEOUT_DEFAULT = 0;
+
+    public static final String RETRY_DELAY_MAX_MS = "retry.delay.max.ms";
+    public static final String RETRY_DELAY_MAX_MS_DOC = "The maximum duration between two consecutive retries (in milliseconds).";
+    public static final long RETRY_DELAY_MAX_MS_DEFAULT = 60000;
+
+    public static final long RETRIES_DELAY_MIN_MS = 300;
+
+    public static final String TOLERANCE_LIMIT = "allowed.max";
+    public static final String TOLERANCE_LIMIT_DOC = "Fail the task if we exceed specified number of errors overall.";
+    public static final String TOLERANCE_LIMIT_DEFAULT = "none";
+
+    public static final RetryWithToleranceOperator NOOP_OPERATOR = new RetryWithToleranceOperator();
+    static {
+        NOOP_OPERATOR.configure(Collections.emptyMap());
+        NOOP_OPERATOR.setMetrics(new ErrorHandlingMetrics());
+    }
+
+    private static final Map<Stage, Class<? extends Exception>> TOLERABLE_EXCEPTIONS = new HashMap<>();
+    static {
+        TOLERABLE_EXCEPTIONS.put(Stage.TRANSFORMATION, Exception.class);
+        TOLERABLE_EXCEPTIONS.put(Stage.HEADER_CONVERTER, Exception.class);
+        TOLERABLE_EXCEPTIONS.put(Stage.KEY_CONVERTER, Exception.class);
+        TOLERABLE_EXCEPTIONS.put(Stage.VALUE_CONVERTER, Exception.class);
+    }
+
+    private long totalFailures = 0;
+    private final Time time;
+    private RetryWithToleranceOperatorConfig config;
+    private ErrorHandlingMetrics errorHandlingMetrics;
+
+    protected ProcessingContext context = new ProcessingContext();
+
+    public RetryWithToleranceOperator() {
+        this(new SystemTime());
+    }
+
+    // Visible for testing
+    public RetryWithToleranceOperator(Time time) {
+        this.time = time;
+    }
+
+    static ConfigDef getConfigDef() {
+        return new ConfigDef()
+                .define(RETRY_TIMEOUT, ConfigDef.Type.LONG, RETRY_TIMEOUT_DEFAULT, ConfigDef.Importance.HIGH, RETRY_TIMEOUT_DOC)
+                .define(RETRY_DELAY_MAX_MS, ConfigDef.Type.LONG, RETRY_DELAY_MAX_MS_DEFAULT, atLeast(1), ConfigDef.Importance.MEDIUM, RETRY_DELAY_MAX_MS_DOC)
+                .define(TOLERANCE_LIMIT, ConfigDef.Type.STRING, TOLERANCE_LIMIT_DEFAULT, in("none", "all"), ConfigDef.Importance.HIGH, TOLERANCE_LIMIT_DOC);
+    }
+
+    /**
+     * Execute the recoverable operation. If the operation is already in a failed state, then simply return
+     * with the existing failure.
+     *
+     * @param operation the recoverable operation
+     * @param <V> return type of the result of the operation.
+     * @return result of the operation
+     */
+    public <V> V execute(Operation<V> operation, Stage stage, Class<?> executingClass) {
```

Comment:
Have you thought about rearranging the parameters so that the operation is last? It's the only parameter that might be a lambda, and having it last would be more readable.

Reply:
In this case everything is quite readable since all the things we're delegating to are super short method calls, I found the code that invokes this quite readable (but of course that's subjective)

Reply:
yeah, I would prefer to keep the operation as the first argument, since it's the most important one. Generally, prefer to order arguments by importance. Also, I found it more readable too.